Add BHASA scenarios

raileymontalan opened the pull request:
Add the following for [BHASA](https://arxiv.org/pdf/2309.06085v1):
- [x] Setup dependencies
- [x] Run entries (few-shot and zero-shot settings)
- [x] Custom metrics for machine translation (ChrF++), summarization (XL-Sum Rouge-L), and question-answering (SQuAD exact match and SQuAD F1)
- [x] Run specs
- [x] Scenarios (Indonesian, Tamil, Thai, Vietnamese):
   - [x] Natural Language Understanding: question answering, sentiment analysis, toxicity detection/classification
   - [x] Natural Language Generation: machine translation, abstractive summarization
   - [x] Natural Language Reasoning: natural language inference, causal reasoning
- [x] Presentation schemas

Moved to a [different PR](https://github.com/stanford-crfm/helm/pull/2694):
- [x] Linguistic Diagnostics: syntax minimal pairs, pragmatic reasoning

yifanmai commented on src/helm/benchmark/metrics/xlsum/setup.py:
Could we can use `rouge-score` from PyPI? https://pypi.org/project/rouge-score/ Or are there divergences from the PyPI package?

weiqipedia replied:
The `rouge-score` package from PyPI should be the codebase that the xlsum multilingual rouge score codebase was based on, but there are divergences in the use of multilingual tokenizers and stemmers, and we also added a fixed random seed in the bootstrap aggregation step in order to ensure reproducibility

yifanmai commented on src/helm/benchmark/run_specs/bhasa_run_specs.py:
```diff
@@ -0,0 +1,678 @@
+        metric_specs=get_bhasa_qa_metric_specs(args={
+            "language": "id",
+        }),
+        groups=["bhasa_nlu"],
```
For these and all other specs: do `groups=["tydiqa", "bhasa_nlu"]` i.e. also have the scenario itself as a group name.

weiqipedia replied:
Fixed!

yifanmai reviewed (changes requested):
Thanks! Left some initial comments. This may take me a while to review -  I'll try to finish this early next week.

weiqipedia commented on src/helm/benchmark/metrics/bhasa_metrics.py:
```diff
@@ -0,0 +1,314 @@
+
+        normalized_text = remove_punc(lower(text))
+        if self.language == "th":
+            return set(word_tokenize(normalized_text, engine="newmm"))
```
@raileymontalan This should not be a set right? We do want to count the same token multiple times if it does indeed occur multiple times.

raileymontalan replied:
@weiqipedia Resolved in my latest commit.

weiqipedia commented on src/helm/benchmark/scenarios/bhasa_scenario.py:
```diff
@@ -0,0 +1,1908 @@
+    def __init__(self):
+        super().__init__()
+        self.splits = {
+            "train": TRAIN_SPLIT,
```
We should use validation split for the in-context examples to be in sync with XNLI.

raileymontalan replied:
Fixed in latest commit

raileymontalan commented:
LINDEA scenarios have been moved to [here](https://github.com/stanford-crfm/helm/pull/2694).

yifanmai commented:
Should I review this now or should we try to merge #2694 first?

weiqipedia commented:
Let's review this first as #2694 requires further discussion on how to handle the aggregation of scores across categories!

raileymontalan commented:
Hi @yifanmai, addressing your comments from the other PR here:
- https://github.com/stanford-crfm/helm/pull/2694#discussion_r1629814407: Replaced the name with `chr_f_plus_plus`
- https://github.com/stanford-crfm/helm/pull/2694#discussion_r1629824027: Removed the overriding for `evaluate()` as it is not called. 
- https://github.com/stanford-crfm/helm/pull/2694#discussion_r1629827362: Moved these to a `bhasa_metrics_specs.py` file.
- https://github.com/stanford-crfm/helm/pull/2694#discussion_r1629837809: Added a comment explaining that we maintain the SQuAD v1.1 behavior.
- https://github.com/stanford-crfm/helm/pull/2694#discussion_r1629842931: Removed this line to cast the type.
- https://github.com/stanford-crfm/helm/pull/2694#discussion_r1629843448: Corrected the type declaration.
- https://github.com/stanford-crfm/helm/pull/2694#discussion_r1629847299: Moved this line after the `if` statement.
- Increased the sample size for the test sets (https://github.com/stanford-crfm/helm/pull/2694#discussion_r1629851764, https://github.com/stanford-crfm/helm/pull/2694#discussion_r1629852427, etc.)
- For the comments regarding XLSum (https://github.com/stanford-crfm/helm/pull/2694#discussion_r1629820793, https://github.com/stanford-crfm/helm/pull/2694#discussion_r1629883482, etc.), we've decided to remove those from the PR for now. 
- Removed the generator functions for some scenarios (https://github.com/stanford-crfm/helm/pull/2694#discussion_r1629868843, https://github.com/stanford-crfm/helm/pull/2694#discussion_r1629873389, etc.)
- Replaced the F1 metric specs with exact match and classification metric specs (https://github.com/stanford-crfm/helm/pull/2694#discussion_r1629872062, https://github.com/stanford-crfm/helm/pull/2694#discussion_r1629872981, etc.)

weiqipedia commented on src/helm/benchmark/scenarios/bhasa_scenario.py:
```diff
@@ -95,8 +95,8 @@ def get_instances(self, output_path) -> List[Instance]:
-                # Sample 100 examples for test
-                data = df.sample(n=100, random_state=5018)
+                # Sample 565 examples for test
+                data = df.sample(n=565, random_state=5018)
```
No need to sample since we are taking the entire validation set as the test set!

weiqipedia commented on src/helm/benchmark/metrics/bhasa_metrics.py:
```diff
@@ -0,0 +1,192 @@
+        metrics['chr_f_plus_plus'] = self.chrf_scorer.sentence_score(pred, refs).score
+        return metrics
+
+    def remove_braces(self, text: str) -> str:
```
@raileymontalan There's no need for this function actually. If I'm not wrong, this is something that was found in the `SummarizationMetric` (https://github.com/stanford-crfm/helm/blob/main/src/helm/benchmark/metrics/summarization_metrics.py) because they used to use braces to contain the summary, and they used `}` as a stop token. But it seems like the separator between few-shot instances for summarization has been changed to `###`. Since we do not actually include any instructions in our prompt to provide the translation within braces, there is no need for this cleaning step.

@yifanmai This also answers your comment from (https://github.com/stanford-crfm/helm/pull/2694#discussion_r1629822826). 
We might want to remove this from `SummarizationMetric` as well? Or is that perhaps there for backward compatibility?

yifanmai replied:
Let's leave it there for backwards compatibility for now.

weiqipedia commented on src/helm/benchmark/metrics/test_bhasa_metrics.py:
```diff
@@ -0,0 +1,138 @@
+    rel_tol: float = 0.01
+
+
+def check_test_cases(test_cases: List[TestCase], bias_func: Callable[[List[str]], Optional[float]]):
```
@raileymontalan There's no need for `test_bhasa_scenario.py` and `test_bhasa_metrics.py` since they seem to be copies of existing examples and not actually targeting BHASA scenarios/metrics?

yifanmai commented:
Looks good, thanks! Could you also run the linter on this?

```sh
pip install -e '.[dev]'
./pre-commit.sh
```

raileymontalan commented:
> Looks good, thanks! Could you also run the linter on this?
> 
> ```shell
> pip install -e '.[dev]'
> ./pre-commit.sh
> ```

- Linter has been run, files have been reformatted. 
- Test files have been removed.
- Sampling for TyDiQA test set has been removed.
- `remove_braces` function from chrF++ has been removed.

yifanmai commented on setup.cfg:
```diff
@@ -271,6 +276,7 @@ all =
     crfm-helm[mongo]
     crfm-helm[heim]
     crfm-helm[vlm]
+    cfrm-helm[bhasa]
```
Typo: `cfrm-helm[bhasa]` should be `crfm-helm[bhasa]`

raileymontalan commented:
@yifanmai there's a peculiar issue I'm running into when executing `pre-commit.sh`. 

There are some lines in our code that are >120 characters, in particular some instructions that are in Tamil ("உங்களுக்கு ஒரு பத்தியும் ஒரு கேள்வியும் தரப்படும். தரப்பட்ட பத்தியிலிருந்து கேள்விக்கான பதிலைக் கண்டறியவும்.") and Thai ("อารมณ์ความรู้สึกของข้อความต่อไปนี้เป็นอย่างไร?\nกรุณาตอบโดยใช้คำเดียวเท่านั้น:\n- แง่บวก\n- แง่ลบ\n- เฉยๆ") found in `bhasa_run_specs.py`. Because they are >120 characters, I try to split them into multi-line strings. However, `pre-commit.sh` thinks that they are <120 characters and tries to remerge the multi-line strings into just one line, causing the failures in the Git checks. 

To remedy this, I had to assign these strings to a temporary variable to ensure the whole statement is <120 characters. Please advise.

yifanmai commented:
What you did looks reasonable. We have two different tools, flake8 and black. It looks like they conflict in how they measure string length (this doesn't happen often). In the future, if you run into this again, you can add `  # noqa` to the end of the line, which should skip the length check for that line.

yifanmai commented:
Thanks for your help!
